File: DevExpress VCL/ExpressPrinting System 4/Demos/CBuilder/EQGrid6RL/EQGrid6RLMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "EQGrid6RLMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxEditRepositoryItems"
#pragma link "dxPSCore"
#pragma link "dxPScxGrid6Lnk"
#pragma link "dxBkgnd"
#pragma link "dxPrnDev"
#pragma link "dxPrnPg"
#pragma link "dxPSCompsProvider"
#pragma link "dxPSEdgePatterns"
#pragma link "dxPSEngn"
#pragma link "dxPSFillPatterns"
#pragma link "dxPSGlbl"
#pragma link "dxPSUtl"
#pragma link "dxWrap"
#pragma link "cxGridCustomView"
#pragma link "cxDataStorage"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxDrawTextUtils"
#pragma link "DemoBasicMain"
#pragma link "dxPScxEditorProducers"
#pragma link "dxPScxExtEditorProducers"
#pragma link "dxPScxPageControlProducer"
#pragma link "dxPSPDFExport"
#pragma link "dxPSPDFExportCore"
#pragma link "dxPSPrVwStd"
#pragma link "dxPScxCommon"
#pragma resource "*.dfm"
TEQGrid6RLMainForm *EQGrid6RLMainForm;
//---------------------------------------------------------------------------
__fastcall TEQGrid6RLMainForm::TEQGrid6RLMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actLookAndFeelKindUltraFlatExecute(TObject *Sender)
{
  if(cxGrid->LookAndFeel->Kind != lfUltraFlat){
    cxGrid->LookAndFeel->Kind = lfUltraFlat;
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actLookAndFeelKindFlatExecute(TObject *Sender)
{
  if(cxGrid->LookAndFeel->Kind != lfFlat){
    cxGrid->LookAndFeel->Kind = lfFlat;
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actLookAndFeelKindStandardExecute(TObject *Sender)
{
  if(cxGrid->LookAndFeel->Kind != lfStandard){
    cxGrid->LookAndFeel->Kind = lfStandard;
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::miNativeStyleClick(TObject *Sender)
{
  cxGrid->LookAndFeel->NativeStyle = !cxGrid->LookAndFeel->NativeStyle;
  ((TCustomAction*)Sender)->Checked = !((TCustomAction*)Sender)->Checked;
}
//---------------------------------------------------------------------------



void __fastcall TEQGrid6RLMainForm::actShowDemoDescriptionExecute(
      TObject *Sender)
{
  lbDescrip->Visible = !lbDescrip->Visible;
  ((TCustomAction*)Sender)->Checked = !((TCustomAction*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actGridNativeStyleExecute(
      TObject *Sender)
{
  cxGrid->LookAndFeel->NativeStyle = !cxGrid->LookAndFeel->NativeStyle;
  ((TCustomAction*)Sender)->Checked = !((TCustomAction*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actDownloadsExecute(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com/downloads/index.asp", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actForumExecute(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://24.234.251.34/dxforum/dxforumisapi.dll/", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actDXOnTheWebExecute(
      TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com/index.shtm", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actProductsExecute(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com/products/index.asp", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::AlwaysEnabled(TObject *Sender)
{
  ((TCustomAction*)Sender)->Enabled = true;
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actExitExecute(TObject *Sender)
{
  Close();  
}
//---------------------------------------------------------------------------


void __fastcall TEQGrid6RLMainForm::CustomizeColumns()
{
  const cDistance = 3;
  const cPeriod = 4;
  const cRadius = 7;
  DecimalSeparator = '.';
  for(int i=0; i < tvPlanets->ColumnCount; i++)
    if((i == cDistance) || (i == cRadius))
      tvPlanets->Columns[i]->DataBinding->ValueTypeClass = __classid(TcxIntegerValueType);
    else
      if (i == cPeriod)
        tvPlanets->Columns[i]->DataBinding->ValueTypeClass = __classid(TcxFloatValueType);
      else
        tvPlanets->Columns[i]->DataBinding->ValueTypeClass = __classid(TcxStringValueType);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::InitRecord(String const Str, int AInt, TStringList* AValues)
{
  Variant V;
  AValues->CommaText = Str;
  for(int i=0; i < AValues->Count; i++)
   if (AValues->Strings[i] != "-") {
     V = AValues->Strings[i];
      if (!VarIsNull(V)) {
        if (tvPlanets->Columns[i]->DataBinding->ValueTypeClass == __classid(TcxFloatValueType))
          V = VarAsType(StrToFloat(V), varDouble);
        tvPlanets->DataController->Values[AInt][i] = V;
      }
   }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::LoadData()
{
  const String AFileName = "nineplanets.txt";
  const int AHeaderLineCount = 2;

  if (!FileExists(AFileName))
    throw Exception("Data file not found");

  TStringList* ARecords = new TStringList();
  TStringList* AValues = new TStringList();
  try {
    ARecords->LoadFromFile(AFileName);
    tvPlanets->BeginUpdate();
    tvPlanets->DataController->RecordCount = ARecords->Count - AHeaderLineCount;
    for (int i=0; i < ARecords->Count - (AHeaderLineCount + 1) + 1; i++)
      InitRecord(ARecords->Strings[i + AHeaderLineCount], i, AValues);
  }
  __finally {
    tvPlanets->EndUpdate();
    delete ARecords;
    delete AValues;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::FormCreate(TObject *Sender)
{
  TDemoBasicMainForm::FormCreate(Sender);
  CustomizeColumns();
  LoadData();
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::FormShow(TObject *Sender)
{
  tvPlanets->DataController->Groups->FullExpand();
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actFullExpandExecute(TObject *Sender)
{
  tvPlanets->DataController->Groups->FullExpand();
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid6RLMainForm::actFullCollapseExecute(TObject *Sender)
{
  tvPlanets->DataController->Groups->FullCollapse();
}
//---------------------------------------------------------------------------

 

File: DevExpress VCL/ExpressPrinting System 4/Demos/Delphi/EQGrid6RL/EQGrid6RLMain.pas
unit EQGrid6RLMain;

{$I ..\cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, dxPSGlbl, dxPSUtl, dxPSEngn, dxPrnPg, dxBkgnd,
  dxWrap, dxPrnDev, dxPSCompsProvider, dxPSFillPatterns, dxPSEdgePatterns,
  dxPSPDFExportCore, dxPSPDFExport, cxDrawTextUtils, dxPSPrVwStd,
  dxPScxEditorProducers, dxPScxExtEditorProducers,
  dxPScxPageControlProducer, dxPSCore, ActnList, ImgList, Menus, ComCtrls,
  ToolWin, StdCtrls, cxGraphics, cxControls, cxLookAndFeels,
  cxLookAndFeelPainters, cxStyles, cxCustomData, cxFilter, cxData,
  cxDataStorage, cxEdit, cxGridLevel, cxGridCustomTableView,
  cxGridTableView, cxGridCustomView, cxGrid, cxClasses,
  cxEditRepositoryItems, dxPScxCommon, dxPScxGrid6Lnk;

type
  TEQGrid6RLMainForm = class(TDemoBasicMainForm)
    edrepMain: TcxEditRepository;
    edrepCenterText: TcxEditRepositoryTextItem;
    edrepRightText: TcxEditRepositoryTextItem;
    StyleRepository: TcxStyleRepository;
    cxStyle1: TcxStyle;
    cxStyle2: TcxStyle;
    cxStyle3: TcxStyle;
    cxStyle4: TcxStyle;
    cxStyle5: TcxStyle;
    cxStyle6: TcxStyle;
    cxStyle7: TcxStyle;
    cxStyle8: TcxStyle;
    cxStyle9: TcxStyle;
    cxStyle10: TcxStyle;
    cxStyle11: TcxStyle;
    cxStyle12: TcxStyle;
    cxStyle13: TcxStyle;
    cxStyle14: TcxStyle;
    tvssDevExpress: TcxGridTableViewStyleSheet;
    cxGrid: TcxGrid;
    tvPlanets: TcxGridTableView;
    tvPlanetsNAME: TcxGridColumn;
    tvPlanetsNO: TcxGridColumn;
    tvPlanetsORBITS: TcxGridColumn;
    tvPlanetsDISTANCE: TcxGridColumn;
    tvPlanetsPERIOD: TcxGridColumn;
    tvPlanetsDISCOVERER: TcxGridColumn;
    tvPlanetsDATE: TcxGridColumn;
    tvPlanetsRADIUS: TcxGridColumn;
    lvPlanets: TcxGridLevel;
    dxComponentPrinterLink1: TdxGridReportLink;
    procedure actFullExpandExecute(Sender: TObject);
    procedure actFullCollapseExecute(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormShow(Sender: TObject);
  private
    procedure CustomizeColumns;
    procedure LoadData;
  end;

var
  EQGrid6RLMainForm: TEQGrid6RLMainForm;

implementation

{$R *.dfm}

procedure TEQGrid6RLMainForm.actFullExpandExecute(Sender: TObject);
begin
  inherited;
  tvPlanets.DataController.Groups.FullExpand;
end;

procedure TEQGrid6RLMainForm.actFullCollapseExecute(Sender: TObject);
begin
  inherited;
  tvPlanets.DataController.Groups.FullCollapse;
end;

procedure TEQGrid6RLMainForm.CustomizeColumns;
const
  cDistance = 3;
  cPeriod = 4;
  cRadius = 7;
var
  I: Integer;
begin
  DecimalSeparator := '.';
  with tvPlanets do
  for I := 0 to ColumnCount - 1 do
    if I in [cDistance, cRadius] then
      Columns[I].DataBinding.ValueTypeClass := TcxIntegerValueType
    else
      if I in [cPeriod] then
      Columns[I].DataBinding.ValueTypeClass := TcxFloatValueType
      else
       Columns[I].DataBinding.ValueTypeClass := TcxStringValueType;
end;

procedure TEQGrid6RLMainForm.LoadData;
const
  AFileName = 'nineplanets.txt';
  AHeaderLineCount = 2;

var
  ARecords, AValues: TStringList;
  I: Integer;

  procedure InitRecord(const Str: string);
  var
    J: Integer;
    V: Variant;
  begin
    AValues.CommaText := Str;
    for J := 0 to AValues.Count - 1 do
     if AValues.Strings[J] <> '-' then
     begin
      V := AValues.Strings[J];
      if not VarIsNull(V) then
      begin
        if tvPlanets.Columns[J].DataBinding.ValueTypeClass = TcxFloatValueType then
          V := StrToFloat(V);
        tvPlanets.DataController.Values[I, J] := V;
      end;
     end;
  end;

begin
  if not FileExists(AFileName) then
    raise Exception.Create('Data file not found');

  ARecords := TStringList.Create;
  AValues := TStringList.Create;

  with ARecords do
  try
    LoadFromFile(AFileName);
    tvPlanets.BeginUpdate;
    tvPlanets.DataController.RecordCount := Count - AHeaderLineCount;
    for I := 0 to Count - (AHeaderLineCount + 1) do
      InitRecord(Strings[I + AHeaderLineCount]);
  finally
    tvPlanets.EndUpdate;
    ARecords.Free;
    AValues.Free;
  end;
end;

procedure TEQGrid6RLMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  CustomizeColumns;
  LoadData;
end;

procedure TEQGrid6RLMainForm.FormShow(Sender: TObject);
begin
  tvPlanets.DataController.Groups.FullExpand;
end;

end.
